### ISSUE TITLE
improvement to distance filter

### ISSUE THREAD

**Issue by socstur** [tananaev/traccar#3462]
Hi Anton,
               I'm seeing devices that are stationary but traccar shows them having motion = yes. I believe whats causing that is the distance filter configuration which filters out the data when the device actually comes to a stop (ie speed~0) since the coordinates are within the distance filter.

To explain more...when a device reports position right before a stop it'll have a speed not equal to 0 (sometimes I see speeds as high as 5 mph). When the device reports its next position which could be several second later ( no less than 10s in my case) it'll have a speed~0 but the coordinates will be the same as before or within the distance filter.  This position and any other subsequent position gets filtered out by the distance filter so traccar never gets a position data where speed~0 even though the device has stopped. This will affect any functions that uses the device motion state for calculations.

Two solutions come to mind..one is to increase the speed threshold. In my case I'll probably have to put it close to 10mph which is not ideal. Another solution is to allow the distance filter to only come into effect after certain number of reported positions are within the range of filtering. Lets say we can let the first two or three positions within the distance filter to not get filtered out. 

what are your thought on this?

**Comment by tananaev** [comment]
You should probably explain what the problem is. What functions does it affect in your case?

**Comment by socstur** [comment]
As an example I have implemented a function that only requests for geocoding when the device is stationary(ie when motion = no ). I can imagine any future functions that uses motion will have similar issues if motion is reported as yes when the device is actually stationary.

**Comment by tananaev** [comment]
I would need an example from official version of Traccar.

**Comment by socstur** [comment]
I cannot think of an official traccar function besides that the fact that if you use the distance filter the "motion" state of the device will be incorrect sometimes. I hope atleast you follow why that is the case. Maybe it might not need resolution but but keep in mind for the future if the motion attribute gets more use.

Do you mind pointing me to where the distance filter is implemented? I'll like to do a simple modification to my version of code where it looks at the previous position data to make sure there was no motion before it implements the distance filter.

**Comment by tananaev** [comment]
It's in `FilterHandler`.

**Comment by socstur** [comment]
Thanks! I made this modification below. It basically makes sure the device has come to a stop before it starts filtering the position. I don't know if its of any value to be included in the official traccar code.

    private boolean filterDistance(Position position, Position last) {
        if (filterDistance != 0 && last != null && !last.getBoolean(last.KEY_MOTION) ) {
            return position.getDouble(Position.KEY_DISTANCE) < filterDistance;
        }
        return false;
    }

**Comment by tananaev** [comment]
Thanks for sharing the code. We will consider including something like that into official version.

**Comment by mongaru** [comment]
Regarding the distance filter I have a real scenario that I am having trouble with.
Scenario: let's say the vehicle stops when the driver gets home at 10 pm. With the distance filter set to any value lets say "50m", no record will be considered since the vehicule is parked. In the morning I check the system but it says that the vehicle haven't been reporting for the last 7 or 8 hours.
Is there a way to update the last connection or maybe accept one record after 1 hour, so it doesn't seem that there is something wrong with the device?

**Comment by tananaev** [comment]
Yes, the is a configuration parameter specifically for that purpose. Read official documentation.

**Comment by mongaru** [comment]
Thanks for the reply Anton. Knowing that I found it in the documentation. I am assuming filter.skipLimit is the property.

**Comment by tananaev** [comment]
Yes, that's the parameter.

**Comment by CarlosBrazil** [comment]
> Thanks! I made this modification below. It basically makes sure the device has come to a stop before it starts filtering the position. I don't know if its of any value to be included in the official traccar code.
> 
> ```java
>     private boolean filterDistance(Position position, Position last) {
>         if (filterDistance != 0 && last != null && !last.getBoolean(last.KEY_MOTION) ) {
>             return position.getDouble(Position.KEY_DISTANCE) < filterDistance;
>         }
>         return false;
>     }
> ```

Hi there,  I have had the same problem like socstur,  and I have used the code that he shared here,  and thanks,  it woked fine,  but is there a way to get this code work with ignition status ?  

like this: 

    if (filterDistance != 0 && last != null && !last.getBoolean(last.KEY_MOTION) && **ignition == false** )   

is that possible ?    if yes,  how to do it the correct way ?

thanks
